### ISSUE TITLE
Geometry Questions

### ISSUE THREAD

**Issue by yjhong89** [xiyichen/morphablediffusion#8]
Hi! 

I am currently fine-tuning pretrained morphable diffusion model (provided) with my custom multi-view face images rendered via blender.

1. How did you acquire facescape_tacking_flame meshes ?? I also possess facescape raw dataset but they only provide raw mesh (ply) and TU-model (their bilinear model) mesh. Pointed out here (mesh / on_mesh)
- https://github.com/xiyichen/morphablediffusion/blob/1daf1280813fad45f9f16f101d772fc51811836e/preprocessing/facescape/process_dataset.py#L111

2. Geometry difference between facescape_tracking_flame mesh and MICA fitted one
- When I display these two meshes on Meshlab, it shows their geometry coordinates are different. (Looking down one is facescape_tracking_flame)

![Image](https://github.com/user-attachments/assets/6a3769cb-75f4-4be2-a66c-1c4ecc6d7c64)

- So there is a code for modifying this. How get this hard-coded values?? https://github.com/xiyichen/morphablediffusion/blob/1daf1280813fad45f9f16f101d772fc51811836e/generate_face.py#L205

3. What is proper size of focal length ?
- My custom dataset is composed of raw mesh and flame wrapped one. And FLAME wrapped mesh follows FLAME template mesh topology and geometry so it seems my geometry seems right if I undo this one. https://github.com/xiyichen/morphablediffusion/blob/1daf1280813fad45f9f16f101d772fc51811836e/ldm/data/facescape.py#L130
- When I render with blender, images have quite low focal length (384.6) compared to this values https://github.com/xiyichen/morphablediffusion/blob/1daf1280813fad45f9f16f101d772fc51811836e/generate_face.py#L166
- Inference result of my fine-tuning result is quite bad. Leftmost image is rendered image I used for fine-tuning. I guess it is due to in-consistency of intrinsic matrix and any advice for this ??

![Image](https://github.com/user-attachments/assets/01a83cd6-3eb2-4876-ab1a-d6978931e39b)

**Comment by xiyichen** [comment]
I fitted the flame meshes with Facescape RGB images and ground-truth camera matrices, using the code in this repo: https://github.com/YuelangX/Multiview-3DMM-Fitting. The fitted flame meshes are thus correspoinding to the real-world camera coordinate systems for the Facescape dataset.

https://github.com/xiyichen/morphablediffusion/blob/1daf1280813fad45f9f16f101d772fc51811836e/generate_face.py#L206 this line is hard-coded to align MICA meshes with fitted Facescape ones. I manually calibrated the two source of meshes in Blender to get those hard-coded values for rotation and translation. Since your data is different, I recommend you to align your flame meshes with the ones in facescape_flame_tracking.zip, if you decide to finetune the pretrained model. If you train from scratch, then you have the freedom to re-do the normalization and don't need to align with the facescape meshes.

For focal length, we use the ground-truth intrinsic from facescape during training. During inference, we always use a similar focal length as the ones in the training data. You can see an example heuristics we use for inference trajectory here: https://github.com/xiyichen/morphablediffusion/blob/1daf1280813fad45f9f16f101d772fc51811836e/generate_face.py#L166. If your focal lengths are very different from this one, then it will fail unless you finetune the model with your data. This is a limitation inhereted from the SyncDreamer model that we based on.

**Comment by yjhong89** [comment]
Thanks for quick response!

- So focal length used in intrinsic (around 1545.23) is similar those with ground-truth intrinsic obtained in here ?? (intrinsic after crop)
https://github.com/xiyichen/morphablediffusion/blob/1daf1280813fad45f9f16f101d772fc51811836e/preprocessing/facescape/process_dataset.py#L222

- As you said, intrinsic difference between training (intrinsic of my custom dataset (around 150~160)) and test may fail inference result. So I try to render with large focal length (576 (width) * focal length (96 mm) / sensor size (32mm) = 1728), does not contain whole image. (Camera position is azimuth/elevation=0, radius 4.5)

![Image](https://github.com/user-attachments/assets/0dc3e43f-2644-4c6e-aa1d-7cda6f439825)

- So I think it would be better to train from scratch including facescape blender rendered images. When you train from scratch, did you train "real" scratch (random initialization) or from other checkpoint like SyncDreamer ?

**Comment by xiyichen** [comment]
Yes. I think the focal length I chose is similar as the ones in facescape after crop. Could you make your face mesh smaller or increase the camera radius during rendering? That could possibly make it fit into the frame.

To train from scratch, I didn't mean to train the entire network from scratch. The clip encoder and the denoising UNet in the diffusion should be initialized from the pretrained models. I trained the entire 3D conditioning branch from scratch, except  the target_encoder https://github.com/xiyichen/morphablediffusion/blob/1daf1280813fad45f9f16f101d772fc51811836e/ldm/models/diffusion/morphable_diffusion.py#L159, which increases the number of channels of the image latent codes from 4 to 16 for volume construction. For this, I loaded the pretrained weights from SyncDreamer. You can check the config and this part of the code https://github.com/xiyichen/morphablediffusion/blob/1daf1280813fad45f9f16f101d772fc51811836e/train_morphable_diffusion.py#L194 to see which modules are loaded from the pretrained models.

**Comment by yjhong89** [comment]
@xiyichen ,
Thanks for detail answer!

Another questions: 
- why did you set radius to 4.5 ??
  - https://github.com/xiyichen/morphablediffusion/blob/1daf1280813fad45f9f16f101d772fc51811836e/generate_face.py#L27
- Why elevation /azimuth fixed to zero?? (aren't they model input??)
  - https://github.com/xiyichen/morphablediffusion/blob/1daf1280813fad45f9f16f101d772fc51811836e/generate_face.py#L158

**Comment by xiyichen** [comment]
https://github.com/xiyichen/morphablediffusion/blob/1daf1280813fad45f9f16f101d772fc51811836e/generate_face.py#L27 this is just a value I chose for the stylegan demos on the project page. I found that the virtual camera trajectory looked quite good with this radius. You could change it based on your needs.

https://github.com/xiyichen/morphablediffusion/blob/1daf1280813fad45f9f16f101d772fc51811836e/generate_face.py#L158 Those are some legacy target elevation and azimuth parameters from SyncDreamer. In their model, those are required as part of the inputs, used in viewpoint embedding https://github.com/liuyuan-pal/SyncDreamer/blob/232a2ec22f0277b4bd429109fb34eaf8f61b13ad/ldm/models/diffusion/sync_dreamer.py#L286. This was helpful in their case for the model to reason about the relative elevation and azimuth difference between input and target views.   In our method, we removed those inputs for better generalization. It works sufficiently well without viewpoint embeddings since we already have the parametric mesh to help with the reconstruction. Those variables are thus unused and were set to 0s. I should have removed them to avoid confusion.

**Comment by yjhong89** [comment]
Thanks for detailed answer!
